src/hotspot/share/opto/vmaskloop.cpp

```diff
@@ -0,0 +1,975 @@
+          return false;
+        }
+      }
+    }
+#ifndef PRODUCT
+    if (TraceMaskedLoop) {
+      tty->print_cr("Nodes in statement [%d] with element type '%s'",
+                    idx, type2name(statement_bottom_type(stmt)));
+      for (uint i = 0; i < stmt->size(); i++) {
+        stmt->at(i)->dump();
+      }
+      tty->cr();
+    }
+#endif
+  }
+
+  return true;
+}
+
+// Analyze loop statements and bail out if any of them is not vectorizable
+bool VectorMaskedLoop::analyze_vectorizability() {
+  if (!find_vector_element_types()) {
+    return false;
+  }
+  if (!vector_nodes_implemented()) {
+    return false;
+  }
+  // Delegate data dependence check to SWPointer utility
+  if (SWPointer::has_potential_dependence(_swptrs)) {
+    trace_msg(nullptr, "Potential data dependence is found in the loop");
+    return false;
+  }
+  if (!analyze_loop_body_nodes()) {
+    return false;
+  }
+  return true;
+}
+
+// Find element basic type for each vectorization candidate node
+bool VectorMaskedLoop::find_vector_element_types() {
+  for (int idx = 0; idx < _stmts.length(); idx++) {
+    Node_List* stmt = _stmts.at(idx);
+    BasicType stmt_bottom_type = statement_bottom_type(stmt);
+    bool subword_stmt = is_subword_type(stmt_bottom_type);
+
+    // Record vector lane size
+    _size_stats.record_size(type2aelembytes(stmt_bottom_type));
+
+    // Set element type for each statement node from bottom to top. Bail out if
+    // the pattern is unsupported
+    for (int i = stmt->size() - 1; i >= 0; i--) {
+      Node* node = stmt->at(i);
+      if (node->is_Mem()) {
+        // Use memory type as its element basic type for memory node
+        BasicType mem_type = node->as_Mem()->memory_type();
+        set_elem_bt(node, mem_type);
+        if (node->is_Load()) {
+          // For load node, check if it has the same vector element size with
+          // the bottom type of the statement
+          if (!same_element_size(mem_type, stmt_bottom_type)) {
+            trace_msg(node, "Vector element size does not match");
+            return false;
+          }
+        }
+      } else {
+        int opc = node->Opcode();
+        if (subword_stmt &&
+            (opc == Op_RShiftI || opc == Op_URShiftI ||
+             opc == Op_AbsI || opc == Op_ReverseBytesI)) {
+          // In any Java arithmetic operation, operands of small integer types
+          // (boolean, byte, char & short) should be promoted to int first. For
+          // some operations, the compiler has to know the operand's higher
+          // order bits, which will be lost in narrowed type. These operations
+          // shouldn't be vectorized if the higher order bits info is unknown.
+          Node* in1 = node->in(1);
+          if (in1->is_Load()) {
+            BasicType mem_type = in1->as_Mem()->memory_type();
+            set_elem_bt(node, mem_type);
+          } else {
+            trace_msg(node, "Subword operand does not have precise type");
```

Comment:
Not clear to me what this means.

Reply:
Precise type info about signedness means that we know exactly whether the data is signed or unsigned. For some operations, such as right shift, results are different for signed and unsigned operands, so C2 has to know the signedness. However, in any Java arithmetic operation, operands of Java subword types are promoted to int first. Sometimes, for example, if an intermediate result is a binary operation of both signed and unsigned, we don't have the precise type info, so we don't know how to vectorize it. (see below example where the signedness info is lost after a short and a char are added)
```
for (int i = 0; i < SIZE; i++) {
  shorts[i] = (shorts[i] + chars[i]) >> 10;
}
```

Reply:
That is annoying. Do you think we can do something about this in the future, or is this just a fundamental restriction of Java / C2?